pkg/virt-handler/vm.go

```diff
@@ -308,11 +308,30 @@ func (d *VirtualMachineController) execute(key string) error {
 		return err
 	}
 
+	log.Log.V(3).Infof("Processing vmi %v, existing: %v\n", vmi.Name, vmiExists)
+	if vmiExists {
+		log.Log.V(3).Infof("vmi is in phase: %v\n", vmi.Status.Phase)
+	}
+
 	domain, domainExists, err := d.getDomainFromCache(key)
 	if err != nil {
 		return err
 	}
 
+	// Ignore domains from an older VMI
+	if vmiExists && domainExists && domain.Spec.Metadata.KubeVirt.UID != vmi.UID {
+		log.Log.Object(vmi).Info("Ignoring domain from an older VMI.")
+		domain = nil
+		domainExists = false
```

Comment:
This is happening now because the domain isn't being cleaned up properly. This case shouldn't occur otherwise. 

Also in the event that this condition is hit where a domain exists for an old VMI, then I think we need to wait for the watchdog file on the old domain to expire and remove the domain before processing the VMI.

Reply:
The problem was: when we stop a VM and restart it immediately, the new VMI could see the old domain. The domain old will still be handled when the old VMI with the matching UID is being processed. Does that make sense?

Reply:
yes, that shouldn't be possible under normal operation.

The reason the old domain still exists in the cache is because of how the domain is being torn down now. This behavior is an artifact of the other changes in this PR.